mixed-content/generic/common.js

```diff
@@ -0,0 +1,238 @@
+}
+
+function queryAudio(url) {
+  queryMedia("audio", {}, {type: "audio/mpeg", src: url});
+}
+
+function queryPicture(url) {
+ var picture = queryMedia("picture", {}, {"srcset": url, "type": "image/png"});
+ createElement("img", {"src": url}, picture);
+}
+
+function queryObject(url) {
+  createElement("object", {"data": url}, document.body);
+}
+
+
+function guid() {
+  function s4() {
+    return (Math.floor((Math.random() + 1.0) * 0x10000))
+           .toString(16).substring(1);
+  }
+  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() +
+         s4() + s4();
+}
+
+function xhrRequest(url, responseType) {
+  return new Promise(function(resolve, reject) {
+    var xhr = new XMLHttpRequest();
+    xhr.open('GET', url, true);
+    xhr.responseType = responseType || "json";
+
+    xhr.onerror = function() {
+      reject(Error("Network Error"));
+    };
+
+    xhr.onload = function() {
+      if (xhr.status != 200)
+        return reject(Error(xhr.statusText));
+
+      resolve(xhr.response);
+    };
+
+    xhr.send();
+  });
+}
+
+function timeoutPromise(timeout) {
+  return new Promise(function(resolve, reject) {
+    setTimeout(resolve, timeout);
+  })
+}
+
+function estimateResponseTime(url, numMeasurements) {
+
+  function measureRoundtripTime(url, index) {
+    return new Promise(function(resolve, reject) {
+      var xhr = new XMLHttpRequest();
+      xhr.open('GET', url);
+      var startTime;
+
+      xhr.onerror = function() {
+        reject(Error("Network Error"));
+      };
+
+      xhr.onload = function() {
+        if (xhr.status != 200)
+          return reject(Error(xhr.statusText));
+
+        var roundtrip_ms = (new Date()).getTime() - startTime;
+        resolve({roundtrip_ms:roundtrip_ms, index:index});
+      };
+
+      startTime = (new Date()).getTime();
+      xhr.send();
+    });
+  }
+
+  function roundTo2Decimals(number) {
+    return Math.round(number * 100.0) / 100.0;
+  }
+
+  function getStats(results) {
+    var stats = {
+      sum: 0,
+      min: 100000,
+      max: -100000,
+      avg: 0,
+      count: 0,
+      stdev: 0,
+    };
+
+    for (var i in results) {
+      stats.sum += results[i];
+      stats.min = Math.min(stats.min, results[i]);
+      stats.max = Math.max(stats.max, results[i]);
+      ++stats.count;
+    }
+
+    // TODO(kristijanburnik): assert count > 0.
+    stats.avg = roundTo2Decimals(stats.sum / stats.count);
+
+    for (var i in results) {
+      var delta = results[i] - stats.avg
+      stats.stdev += delta * delta;
+    }
+
+    stats.stdev /= stats.count;
+    stats.stdev = roundTo2Decimals(Math.sqrt(stats.stdev));
+
+    return stats;
+  }
+
+  return new Promise(function(resolve, reject) {
+    var results = [];
+    var p = measureRoundtripTime(url, 0);
+
+    for (var i = 0; i < numMeasurements - 1; ++i) {
+      p = p.then(function(result) {
+        results.push(result.roundtrip_ms);
+        return measureRoundtripTime(url, result.index + 1);
```

Comment:
Yes, I am still thinking about this approach. I wouldn't however be very satisfied by setting some fixed timeout for the resource... 

I would consider not doing this measurement on every test, however it's more likely the subsequent requests would take approx the same time if they're closer together, I'd say. For example, if the machine running the tests for some (external) reason gets a higher load on a test at index 100 and a response time takes 3x longer than the first 100 tests, the affected test should most likely still work fine as we're taking the response ETA at that busy time chunk.

Currently I only make 3 extra requests per test, which is not that horrible. They are also done in sequence.